---
title: "methylation distance on promoters"
author: "isaclee"
date: "5/6/2019"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(GenomicRanges)
repodir="/home/isac/Code/nanoNOMe"
source(file.path(repodir,"scripts/ggplot_theme.R"))
source(file.path(repodir,"scripts/methylation_plot_utils.R"))
```

## write out the names of high vs low expression genes
```{r rnaseq,include=FALSE,eval=FALSE}
root="/kyber/Data/Nanopore/projects/nanonome/analysis"
setwd(root)
# annotations
tss.fp="data/hg38/hg38_genes.TSS.200bp.bed"
# rnaseq database
exp.fp=c("data/gm12878/GM12878_rnaseq_1.tsv",
         "data/gm12878/GM12878_rnaseq_2.tsv")
#read in tss db
tss.gr=load_db(tss.fp,extracols="name")
tss.gr$id = sapply(strsplit(tss.gr$id,"[.]"),"[[",1)

# expression data parsing
qtiles=seq(0,1,.25)
exp.list=lapply(exp.fp,read_tsv)
filter.list=lapply(seq_along(exp.list),function(i){
    x=exp.list[[i]]
    id=sapply(strsplit(x$gene_id,"[.]"),"[[",1)
    y=tibble(id=id,fpkm=x$FPKM_ci_upper_bound)
    y=y[which(y$id %in% tss.gr$id),]
    unexpressed = tss.gr[-which(tss.gr$id %in% y$id)]
    qs=quantile(y$fpkm,probs=c(0.25,0.5,0.75))
    y = y%>%mutate(qtile=ifelse(fpkm>=qs[1],3,4),
               qtile=ifelse(fpkm>=qs[2],2,qtile),
               qtile=ifelse(fpkm>=qs[3],1,qtile),
               rep=i)
    bind_rows(y,
              tibble(id=unexpressed$id,fpkm=0,qtile=5,rep=i))
})
exp=do.call(rbind,filter.list)
# using only genes that are consistent b/w the two replicates, taking average
exp.qtile=exp%>%select(-fpkm)%>%
    spread(rep,qtile)%>%filter(`1`==`2`)%>%mutate(qtile=`1`)
exp.fpkm=exp[which(exp$id %in% exp.qtile$id),] %>%
    group_by(id,qtile)%>%summarize(fpkm=mean(fpkm))

# merge with tss.gr
tss.gr=tss.gr[which(tss.gr$id %in% exp.fpkm$id)]
tss.gr$qtile=exp.fpkm$qtile[match(tss.gr$id,exp.fpkm$id)]
tss.gr$fpkm=exp.fpkm$fpkm[match(tss.gr$id,exp.fpkm$id)]

## export high and low expression genes
if (TRUE){
    tss.high = tss.gr[which(tss.gr$qtile==1)]
    tss.low = tss.gr[which(tss.gr$qtile==4)]
    tss.no = tss.gr[which(tss.gr$qtile==4)]
    bed.high = GRangesTobed(tss.high)
    bed.low = GRangesTobed(tss.low)
    bed.no = GRangesTobed(tss.no)
    names.high = bed.high[,"name"]
    names.low = bed.low[,"name"]
    names.no = bed.no[,"name"]
    outdir=file.path(root,"data/gm12878")
    highpath = file.path(outdir,"GM12878_rnaseq_highexpression.geneid.txt")
    lowpath = file.path(outdir,"GM12878_rnaseq_lowexpression.geneid.txt")
    nopath = file.path(outdir,"GM12878_rnaseq_noexpression.geneid.txt")
    write_tsv(names.high,highpath,col_names=F)
    write_tsv(names.low,lowpath,col_names=F)
    write_tsv(names.no,lowpath,col_names=F)
}

```
## subset gene windows by expression
```{bash genesubset,include=FALSE,eval=FALSE}
export PATH="/home/isac/Code/miniconda3/bin:$PATH"
root="/kyber/Data/Nanopore/projects/nanonome/analysis"
dir=$root/data/gm12878

regs=$root/data/hg38/hg38_genes.TSS.slop2000bp.bed
highnames=$dir/GM12878_rnaseq_highexpression.geneid.txt
lownames=$dir/GM12878_rnaseq_noexpression.geneid.txt
highregs=$dir/GM12878_highexpression_genes.TSS.2000bp.bed
lowregs=$dir/GM12878_lowexpression_genes.TSS.2000bp.bed

awk 'NR==FNR{c[$1]++;next};c[$7]' $highnames $regs > $highregs
awk 'NR==FNR{c[$1]++;next};c[$7]' $lownames $regs > $lowregs
```

## get distances between open marks
```{bash getdist,include=FALSE, eval=FALSE}
export PATH="/home/isac/Code/miniconda3/bin:$PATH"
root="/kyber/Data/Nanopore/projects/nanonome/analysis"
scr="/home/isac/Code/nanoNOMe/scripts/methylation_distance.py"
outdir=$root/plots/mdist

mbed="$root/data/nanonome/pooled/mbed/GM12878_nanoNOMe.pooled.gpc.meth.bed.gz"

# high
reg="$root/data/gm12878/GM12878_highexpression_genes.TSS.2000bp.bed"
regname="highexpression_promoters"
cell="GM12878"
out=$outdir/$cell.gpc.methdistance.$regname.txt
shuf -n1000 $reg | $scr -i $mbed > $out
# low
reg="$root/data/gm12878/GM12878_noexpression_genes.TSS.2000bp.bed"
regname="noexpression_promoters"
cell="GM12878"
out=$outdir/$cell.gpc.methdistance.$regname.txt
shuf -n1000 $reg | $scr -i $mbed > $out
```

```{r plot,echo=FALSE,message=FALSE}
root="/kyber/Data/Nanopore/projects/nanonome/analysis"
outdir=file.path(root,"plots/mdist")
regnames=c("highexpression_promoters","noexpression_promoters")
cell="GM12878"
tb.list = list()
for (regname in regnames){
    dfp=file.path(outdir,paste(cell,"gpc.methdistance",regname,"txt",sep="."))
    tb = read_tsv(dfp)
    tb.list[[regname]] = tb %>% mutate(regname=regname)
    tb = tb[which(tb$region=="all"),]
    
    g = ggplot(tb,aes(x=distance,y=frequency)) +
      geom_density(stat="identity") +
      lims(x=c(0,1000),y=c(0,10000))+
      ggtitle(regname)
    
    if (length(unique(tb$region)) > 1){
        g = g + facet_wrap(.~region)
    }
    print(g)
}
## overlay distance plots
tb.list = lapply(tb.list,function(x){
  x[which(x$region=="all"),]
})
tb.all = do.call(rbind,tb.list)
g = ggplot(tb.all,aes(x=distance,y=frequency,group=regname,color=regname)) +
  geom_density(stat="identity") +
  lims(x=c(0,1000),y=c(0,10000))
print(g)

## check avg distance?
tb.all = do.call(rbind,tb.list) %>%
  filter(distance<1000)
avgdist = tb.all %>% group_by(regname,region)%>%
  summarize(dist=sum(distance*frequency)/sum(frequency))
avgdist[which(avgdist$region=="all"),]
avgdist %>% 
  summarize(median = median(dist),
            mean = mean(dist))

ggplot(avgdist,aes(x=regname,y=dist))+geom_boxplot()
```